Add user operation E2E tests

matthewwalsh0 opened the pull request:
## **Description**

Add initial E2E tests and associated infrastructure for user operations.

Specifically:

- Add the `bundler` helper which uses the [@metamask/test-bundler](https://github.com/MetaMask/test-bundler) package.
- Patch `ganache` to fix revert reason detection in `@ethersproject`.
- Add `entrypoint`, `simpleAccountFactory`, and `verifyingPaymaster` to `smart-contracts.ts` and `GanacheSeeder`.
- Add the `useBundler` and `usePaymaster` options to `withFixtures` to:
  - Deploy the above 4337 smart contracts.
  - Transfer funds to the sender and fixture account.
  - Deposit funds to paymaster.
  - Start the bundler server.
- Add E2E tests to verify the use of user operations from dApp, send, and swap transactions.
  - The tests and assertions are not exhaustive but aim to provide a baseline for future additions.

## **Related issues**

Fixes: [#1738](https://github.com/MetaMask/MetaMask-planning/issues/1738)

## **Manual testing steps**

_Pending creation of test snap._

## **Screenshots/Recordings**

### **Before**

### **After**

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it is solved.
- [x] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [x] In case it's not yet "ready for review", I've set it to "draft".
  - [x] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

matthewwalsh0 committed:
Add user operation controller

Move transaction methods to separate file.
Add paymaster alert to transaction confirmation.
Add paymaster and user operation selectors.

matthewwalsh0 committed:
Fix polling controller patch

matthewwalsh0 committed:
Fix code fence

Update Lavamoat policies.

matthewwalsh0 committed:
Move selectors to separate file

Add env for entrypoint.
Add selector unit tests.

matthewwalsh0 committed:
Move user operation controller listeners to methods

Add unit test for paymaster alert.
Add message for alert text.

matthewwalsh0 committed:
Add properties to sentry test state

matthewwalsh0 committed:
Add initial E2E tests for user operations

Add bundler E2E helper.
Patch Ganache.
Add 4337 smart contracts to seeder and helper.

matthewwalsh0 committed:
Update Lavamoat policies

matthewwalsh0 committed:
Remove yarn duplication

matthewwalsh0 committed:
Update Lavamoat policies

matthewwalsh0 committed:
Fix code fencing

matthewwalsh0 committed:
Fix failing E2E tests

matthewwalsh0 committed:
Fix linting

matthewwalsh0 committed:
Fix MMI and multichain tests

matthewwalsh0 commented:
@metamaskbot update-policies

metamaskbot committed:
Update LavaMoat policies

matthewwalsh0 commented:
@SocketSecurity ignore-all

matthewwalsh0 committed:
Use updated test bundler

matthewwalsh0 committed:
Assert transaction details using receipt

matthewwalsh0 committed:
Get user operation hash from bundler

matthewwalsh0 committed:
Merge branch 'develop' into feat/user-operations

matthewwalsh0 committed:
Update yarn lock

matthewwalsh0 committed:
Update bundler config

matthewwalsh0 committed:
Merge branch 'develop' into feat/user-operations

matthewwalsh0 committed:
Fix yarn lock

metamaskbot committed:
Update LavaMoat policies

dan437 committed:
New Settings: Show extension in full size view (#22383)

Co-authored-by: Danica Shen <zhaodanica@gmail.com>

matthewwalsh0 committed:
Merge branch 'develop' into feat/user-operations

matthewwalsh0 committed:
Remove yarn duplication

shane-t committed:
chore: disable link to portfolio staking in MMI (#22511)

## **Description**

MMI and MetaMask both have staking features that link to different
places. This hides the MetaMask staking feature in MMI.

## **Related issues**

Fixes:

Jira issue
https://consensyssoftware.atlassian.net/jira/software/projects/MMI/boards/315?selectedIssue=MMI-4426

## **Manual testing steps**

1. Create an MMI build
2. Open the home screen on mainnet
3. Observe that there is no link to stake ether from the assets list.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**


![image](https://github.com/MetaMask/metamask-extension/assets/1504499/c8a41e5b-3edb-4645-9c06-58350a7c8d73)

### **After**

<img width="369" alt="image"
src="https://github.com/MetaMask/metamask-extension/assets/1504499/2488c59d-ae23-4ceb-9a15-cecc8d1c06a5">
## **Pre-merge author checklist**

- [X] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [X] I've clearly explained what problem this PR is solving and how it
is solved.
- [X] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [x] In case it's not yet "ready for review", I've set it to "draft".
- [x] In case it's "ready for review", I've changed it from "draft" to
"non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

metamaskbot committed:
Update LavaMoat policies

Gudahtt committed:
chore: Upgrade `@metamask/announcement-controller` to v5.0.1 (#22499)

## **Description**

The package `@metamask/announcement-controller` has been updated to v5.
The only breaking changes were to types, so this required no changes.

See here for the changelog:
https://github.com/MetaMask/core/blob/main/packages/announcement-controller/CHANGELOG.md#501

## **Related issues**

N/A

## **Manual testing steps**

N/A

## **Screenshots/Recordings**

N/A

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it
is solved.
- [x] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [x] In case it's not yet "ready for review", I've set it to "draft".
- [x] In case it's "ready for review", I've changed it from "draft" to
"non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

---------

Co-authored-by: MetaMask Bot <metamaskbot@users.noreply.github.com>

matthewwalsh0 committed:
Merge branch 'feat/user-operations' into test/user-operations-e2e

HowardBraham committed:
fix: QR/Keystone notification window disappears (#22460)

## **Description**

Fixes the longstanding bug where the notification window disappears when
using a QR wallet.

## **Related issues**

Fixes: #19780

## **Manual testing steps**

1. Use any QR wallet (AirGap Vault is free)
2. Go to https://metamask.github.io/test-dapp/
3. Use any of the Sign methods (useful for testing because they do not
require gas in your account)
4. Click 'Sign' in the Notification window that appears
5. The window **does not** close

HowardBraham committed:
Added launch.json configuration to run E2E tests (#22495)

## **Description**

Adds an easy way to run and debug E2E tests in VSCode by pressing F5

## **Related issues**

Builds upon: #22416

## **Manual testing steps**

1. Open a spec.js or spec.ts file
2. If you want to, set a breakpoint
3. Press F5 to start debugging
4. A menu opens up at the top, choose a browser
5. The current E2E test should run in your chosen browser
6. If you had set a breakpoint, it should stop at the breakpoint

FrederikBolding committed:
Fix snaps custom UI styling issues (#22443)

## **Description**
Fixes a number of issues/regressions in snaps custom UI.
- Fix header sizing issues
- Add missing margins

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**


![image](https://github.com/MetaMask/metamask-extension/assets/1561200/725c6184-0dfd-45f3-93fa-ea4724fbf8ae)

### **After**


![image](https://github.com/MetaMask/metamask-extension/assets/1561200/77034ac8-adac-41f5-9978-2379a99d4d8b)

matthewwalsh0 committed:
Use released package versions

garrettbear committed:
Replace RGBA static values (#22473)

## **Description**
Replace static RGBA values to a token from `design-tokens` that best
matches. If unavailable, a temporary file has been set up for any new
tokens needed but not yet provided by the `design-tokens` package.

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

## **Related issues**

Fixes: #22360 

## **Manual testing steps**

1. Run locally
2. Compare [local
storybook](http://localhost:6006/?path=/docs/getting-started-introduction--docs)
to [production
storybook](https://metamask.github.io/metamask-storybook/?path=/story/getting-started-introduction--docs)
on the components color change
3. Be sure to test light and dark mode

## **Screenshots/Recordings**

See code for before and after photos or video of each change

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it
is solved.
- [x] I've linked related issues
- [ ] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [x] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [x] In case it's not yet "ready for review", I've set it to "draft".
- [x] In case it's "ready for review", I've changed it from "draft" to
"non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

---------

Co-authored-by: George Marshall <george.marshall@consensys.net>

matthewwalsh0 committed:
Fix unit tests

matthewwalsh0 committed:
Fix additional unit tests

matthewwalsh0 committed:
Merge branch 'develop' into feat/user-operations

matthewwalsh0 committed:
Revert base controller versions

matthewwalsh0 committed:
Fix selector

matthewwalsh0 committed:
Merge branch 'develop' into feat/user-operations

metamaskbot committed:
Update LavaMoat policies

matthewwalsh0 committed:
Merge branch 'develop' into feat/user-operations

matthewwalsh0 committed:
Merge branch 'develop' into test/user-operations-e2e

matthewwalsh0 committed:
Update tests for example snap

matthewwalsh0 committed:
Set chain config

Add test template method.
Add paymaster test.

matthewwalsh0 committed:
Deploy paymaster only if needed

matthewwalsh0 committed:
Merge branch 'develop' into test/user-operations-e2e

matthewwalsh0 committed:
Move constants

matthewwalsh0 committed:
Fix pipeline

matthewwalsh0 committed:
Fix unit test

matthewwalsh0 committed:
Merge branch 'develop' into test/user-operations-e2e

matthewwalsh0 committed:
Run tests only on flask

matthewwalsh0 committed:
Update Lavamoat policies

matthewwalsh0 committed:
Remove yarn duplication

matthewwalsh0 commented:
@metamaskbot update-policies

metamaskbot committed:
Update LavaMoat policies

matthewwalsh0 committed:
Fix firefox

matthewwalsh0 committed:
Merge branch 'test/user-operations-e2e' of github.com:MetaMask/metamask-extension into test/user-operations-e2e

matthewwalsh0 committed:
Merge branch 'develop' into test/user-operations-e2e

matthewwalsh0 committed:
Remove patch

matthewwalsh0 committed:
Merge branch 'develop' into test/user-operations-e2e

dbrans commented on test/e2e/helpers.js:
```diff
@@ -43,9 +48,13 @@ async function withFixtures(options, testSuite) {
+
   const fixtureServer = new FixtureServer();
   const ganacheServer = new Ganache();
+  const bundlerServer = new Bundler();
```
Why not just have `initBundler` return the bundler directly? It feels odd to set up `bundlerServer` here without first confirming `useBundler` is actually `true`. Keeps things cleaner and more intuitive, don't you think?

matthewwalsh0 replied:
That's technically not the server itself, but the E2E helper.

I've named it `bundlerServer` and put it there for consistency so it's defined next to the other helpers, specifically fixtures and ganache.

dbrans commented on test/e2e/bundler.ts:
```diff
@@ -0,0 +1,66 @@
+        this.#userOperationHashes.push(userOperationHash);
+      });
+
+      await this.#server.asyncStart();
```
I think `startBundler` already [calls `asyncStart` via `runBundler`](https://github.com/MetaMask/test-bundler/blob/main/src/bundler/runBundler.ts#L250):

matthewwalsh0 replied:
It does, but `asyncStart` is just a reference to a promise variable so we use it to wait for the server to finish it's initialisation checks and be considered ready.

matthewwalsh0 committed:
Convert tests to typescript

Duplicate assertions in each test for clarity.
Move new contracts to separate files.

matthewwalsh0 commented on test/e2e/bundler.ts:
```diff
@@ -0,0 +1,66 @@
+  }
+
+  async getUserOperationReceipt(userOperationHash: string) {
+    const response = await fetch(BUNDLER_URL, {
```
We could, but I didn't want to couple unnecessarily to the test bundler API as it feels more "end to end" to do the raw request ourselves.

matthewwalsh0 commented on test/e2e/flask/user-operations.spec.js:
```diff
@@ -0,0 +1,257 @@
+  await driver.delay(1000);
+
+  // Confirm Connect Modal
+  await switchToNotificationWindow(driver, 3);
```
Removed.

matthewwalsh0 commented on test/e2e/flask/user-operations.spec.js:
```diff
@@ -0,0 +1,257 @@
+async function installExampleSnap(driver) {
+  // Navigate to Site
+  await driver.openNewPage(ERC_4337_ACCOUNT_SNAP_URL);
+  await driver.delay(1000);
```
Completely agree. I think I copied these originally from elsewhere or maybe added them for debug, but the tests pass without so have removed.

matthewwalsh0 commented on test/e2e/flask/user-operations.spec.js:
We've never created TypeScript tests prior as I understood the standard was JavaScript hence why all the petnames tests are just JavaScript for example.

That said, it seems the accounts team have created some TypeScript E2E tests so converted this one.

matthewwalsh0 commented on test/e2e/flask/user-operations.spec.js:
```diff
@@ -0,0 +1,257 @@
+    ])}`,
+  );
+
+  await new Promise((resolve) => setTimeout(resolve, 5000));
```
No idea 😄 Didn't appear to be needed so have removed.

matthewwalsh0 commented on test/e2e/flask/user-operations.spec.js:
```diff
@@ -0,0 +1,257 @@
+}
+
+async function confirmTransaction(driver) {
+  await switchToNotificationWindow(driver, 4);
```
We have 4 because we've opened the snap site at this point. But it's no longer needed so I now close it before continuing.

matthewwalsh0 commented on test/e2e/flask/user-operations.spec.js:
```diff
@@ -0,0 +1,257 @@
+}
+
+async function createSnapAccount(driver, privateKey, salt) {
+  await driver.switchToWindowWithTitle('Account Abstraction Snap');
```
Done.

matthewwalsh0 commented on test/e2e/flask/user-operations.spec.js:
```diff
@@ -0,0 +1,257 @@
+async function expectTransactionDetail(driver, rowIndex, expectedText) {
+  await driver.findElement({
+    css: `[data-testid="transaction-breakdown-row"]:nth-child(${
+      2 + rowIndex
```
I think of it as part of the selector so it's just the format of the HTML and likely header rows as you say.

matthewwalsh0 commented on test/e2e/flask/user-operations.spec.js:
```diff
@@ -0,0 +1,257 @@
+  });
+}
+
+async function expectTransactionDetails(driver, bundlerServer) {
```
Renamed.

matthewwalsh0 commented on test/e2e/flask/user-operations.spec.js:
```diff
@@ -0,0 +1,257 @@
+
+  await expectTransactionDetail(
+    driver,
+    3,
```
Great idea, done.

matthewwalsh0 commented on test/e2e/flask/user-operations.spec.js:
```diff
@@ -0,0 +1,257 @@
+
+describe('User Operations', function () {
+  it('from dApp transaction', async function () {
+    await withAccountSnap({ currentTest: this.test }, async (driver) => {
```
My original concern was just the duplication of assertions hence why it's included in the template.

But I agree those assertions could diverge in future, plus they are just a few lines so have moved out of the template for clarity.

matthewwalsh0 commented on test/e2e/seeder/smart-contracts.js:
```diff
@@ -47,13 +47,1210 @@ const multisigFactory = {
   abi: multisigAbi,
 };
 
```
Added, and moved them to separate files as I never liked them all in one huge file.

matthewwalsh0 commented on test/e2e/seeder/ganache-seeder.js:
```diff
@@ -60,9 +70,44 @@ class GanacheSeeder {
+  async transfer(to, value) {
+    const signer = this.#getSigner();
+
+    const sendAccount = await signer.sendTransaction({
```
Renamed.

matthewwalsh0 commented on test/e2e/helpers.js:
```diff
@@ -1024,6 +1043,33 @@ async function getCleanAppState(driver) {
   );
 }
 
+async function initBundler(bundlerServer, ganacheServer, usePaymaster) {
```
It is in a technical sense, but at a high level it is preparing all the contracts and accounts required to use the bundler. Plus it adds some clarity as it matches the `useBundler` option.

matthewwalsh0 committed:
Merge branch 'develop' into test/user-operations-e2e

dbrans commented on development/build/scripts.js:
```diff
@@ -1223,7 +1223,7 @@ async function setEnvironmentVariables({
     EIP_4337_ENTRYPOINT:
       variables.getMaybe('EIP_4337_ENTRYPOINT') ||
-      '0x5FF137D4b0FDCD49DcA30c7CF57E578a026d2789',
+      (testing ? '0x18b06605539dc02ecD3f7AB314e38eB7c1dA5c9b' : undefined),
```
Should we import this from test/e2e/constants.ts?

matthewwalsh0 replied:
Unfortunately we can't since this isn't part of the extension bundle, but the build system.

matthewwalsh0 committed:
Merge branch 'develop' into test/user-operations-e2e

dbrans reviewed (approved):
Fantastic PR!
